src/io/g15.rs

```diff
@@ -0,0 +1,251 @@
+#![allow(non_upper_case_globals)]
+#![allow(unused_variables)]
+
+use regex::{Regex, Captures};
+use std::fmt;
+use std::fs;
+use std::num::ParseIntError;
+
+#[derive(Debug)]
+pub enum G15BadIndex {
+    IndexOutOfBounds, // Occurs when m_someArray[X] has some X greater than 33 (since these are static 34 element arrays)
+    BadInt,           // Occurs when an int fails to parse into the given unsigned range (u32 or u64)
+}
+
+impl fmt::Display for G15BadIndex {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f,"bad G15 value: {:?}", self)
+    }
+}
+
+impl From<ParseIntError> for G15BadIndex {
+    fn from(value: ParseIntError) -> Self {
+        G15BadIndex::BadInt
+    }
+}
+
+// A RegMatch struct contains a Regex and the corresponding function to add the result to the G15Player vec
+// if the regex results in a match.
+pub struct RegMatch(Regex, fn(caps: Captures, players: &mut Vec<G15Player>)->Result<(), G15BadIndex>);
+impl RegMatch {
+    pub fn new(reg_str: &str, func: fn(caps: Captures, players: &mut Vec<G15Player>)->Result<(), G15BadIndex>) -> RegMatch {
+        let regers = Regex::new(reg_str).unwrap();
+        RegMatch (regers, func)
+    }
+}
+
+// m_iAmmo[1] integer (32)      --match--> (player idx) (ammo count)
+pub const REGEX_I_AMMO: &str = r#"^m_iAmmo\[(\d+)\]\s+integer\s+\((\d+)\)$"#;
+pub fn parse_ammo(caps: Captures, players: &mut Vec<G15Player>) -> Result<(), G15BadIndex> {
+    let idx: usize = caps[1].parse().unwrap();

```

Comment:
You created an error variant for parsing ints but are not using it here for indexes. `.unwrap()` -> `?` should just work here I believe, with how you've got the errors set up. I think the `ParseIntError` is the same between `u32` and `usize`.

Reply:
I didn't bother changing these because I was fairly certain that this capture group will never fail to parse due to the regex and the type, but there is no harm in still doing that, and of course I could always be wrong. Also `?` is shorter than `.unwrap()` so that automatically makes it pretty nice

Reply:
Yeah, I agree it's pretty unlikely to ever fail, but the error handling framework you've set up already makes it super easy to just change the unwraps to ?s, so may as well. The less unwraps the better.